libraries/Microsoft.Bot.Connector/Authentication/Retry.cs

```diff
@@ -0,0 +1,54 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Text;
+using System.Threading.Tasks;
+
+namespace Microsoft.Bot.Connector.Authentication
+{
+    public static class Retry
+    {
+        public static async Task<TResult> Run<TResult>(Func<Task<TResult>> task, Func<Exception, int, RetryParams> retryExceptionHandler)
+        {
+            RetryParams retry = RetryParams.StopRetrying;
+            List<Exception> exceptions = new List<Exception>();
+            int currentRetryCount = 0;
+
+            do
+            {
+                try
+                {
+                    return await task().ConfigureAwait(false);
+                }
+                catch (Exception ex)
+                {
+                    exceptions.Add(ex);
+
+                    retry = retryExceptionHandler(ex, currentRetryCount);
+                }
+
+                if (retry.ShouldRetry)
+                {
+                    currentRetryCount++;
+                    await Task.Delay(retry.RetryAfter.WithRandomNoise()).ConfigureAwait(false);
+                }
+
+            } while (retry.ShouldRetry);
+
+            throw new AggregateException("Failed to acquire token for client credentials.", exceptions);
+        }
+    }
+
+    public static class TimeSpanExtensions
+    {
+        private static Random random = new Random();
+
+        public static TimeSpan WithRandomNoise(this TimeSpan delay, double multiplier = 0.1)
```

Comment:
I'm used to seeing this  called jitter, but noise works just fine. :)

Reply:
Oh I translated from the spanish word for this :) I didn't know Jitter was the right word, will rename!

Reply:
Renamed